akka-stream/src/main/scala/akka/stream/scaladsl/SinkRef.scala

```diff
@@ -0,0 +1,212 @@
+/**
+ * Copyright (C) 2018 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.scaladsl
+
+import akka.Done
+import akka.actor.{ ActorRef, Terminated }
+import akka.event.Logging
+import akka.stream._
+import akka.stream.StreamRefs
+import akka.stream.impl.StreamRefsMaster
+import akka.stream.stage._
+import akka.util.{ OptionVal, PrettyDuration }
+
+import scala.concurrent.{ Future, Promise }
+import scala.util.Try
+
+object SinkRef {
+  def source[T](): Source[T, Future[SinkRef[T]]] =
+    Source.fromGraph(new SourceRef[T](OptionVal.None, canMaterializeSinkRef = true))
+
+  // TODO Implement using TCP
+  // steps:
+  // - lazily, but once bind a port
+  // - advertise to other side that they may send data into this port
+  // -- "passive mode" ftp ;-)
+  // def bulkTransferSource(port: Int = -1): Source[ByteString, SinkRef[ByteString]] = ???
+}
+
+/**
+ * The dual of SourceRef.
+ *
+ * This is the "handed out" side of a SinkRef. It powers a Source on the other side.
+ *
+ * Do not create this instance directly, but use `SinkRef` factories, to run/setup its targetRef.
+ *
+ * We do not materialize the refs back and forth, which is why the 2nd param.
+ */
+final class SinkRef[In] private[akka] (
```

Comment:
I think after the refactoring proposals from https://github.com/ktoso/akka/pull/33/files, we should think about renaming the stages, so it's easier to see which one is the sending and which the receiving side.